pkg/infrastructure/azure/azure.go

```diff
@@ -1,14 +1,341 @@
+	imageVersion := rawImageVersion[:len(rawImageVersion)-6]
+
+	galleryName := fmt.Sprintf("gallery_%s", strings.ReplaceAll(in.InfraID, "-", "_"))
+	galleryImageName := in.InfraID
+	galleryImageVersionName := imageVersion
+	galleryGen2ImageName := fmt.Sprintf("%s-gen2", in.InfraID)
+	galleryGen2ImageVersionName := imageVersion
+
+	headResponse, err := http.Head(imageURL)
+	if err != nil {
+		return fmt.Errorf("failed HEAD request for image URL %s: %w", imageURL, err)
+	}
+
+	imageLength := headResponse.ContentLength
+	if imageLength%512 != 0 {
+		return fmt.Errorf("image length is not alisnged on a 512 byte boundary")
+	}
+
+	tokenCredential := session.TokenCreds
+	storageURL := fmt.Sprintf("https://%s.blob.core.windows.net", storageAccountName)
+	blobURL := fmt.Sprintf("%s/%s/%s", storageURL, containerName, blobName)
+
+	// Create user assigned identity
+	userAssignedIdentityName := fmt.Sprintf("%s-identity", in.InfraID)
+	armmsiClientFactory, err := armmsi.NewClientFactory(subscriptionID, tokenCredential, nil)
+	if err != nil {
+		log.Fatalf("failed to create armmsi client: %v", err)
+	}
+	userAssignedIdentity, err := armmsiClientFactory.NewUserAssignedIdentitiesClient().CreateOrUpdate(
+		ctx,
+		resourceGroupName,
+		userAssignedIdentityName,
+		armmsi.Identity{
+			Location: ptr.To(platform.Region),
+			Tags:     tags,
+		},
+		nil,
+	)
+	if err != nil {
+		log.Fatalf("failed to create user assigned identity %s: %v", userAssignedIdentityName, err)
+	}
+	principalID := *userAssignedIdentity.Properties.PrincipalID
+
+	logrus.Debugf("UserAssignedIdentity.ID=%s", *userAssignedIdentity.ID)
+	logrus.Debugf("PrinciapalID=%s", principalID)
+
+	// Create storage account
+	createStorageAccountOutput, err := CreateStorageAccount(ctx, &CreateStorageAccountInput{
+		SubscriptionID:     subscriptionID,
+		ResourceGroupName:  resourceGroupName,
+		StorageAccountName: storageAccountName,
+		CloudName:          platform.CloudName,
+		Region:             platform.Region,
+		Tags:               tags,
+		TokenCredential:    tokenCredential,
+		CloudConfiguration: cloudConfiguration,
+	})
+	if err != nil {
+		return err
+	}
+
+	storageAccount := createStorageAccountOutput.StorageAccount
+	storageClientFactory := createStorageAccountOutput.StorageClientFactory
+	storageAccountKeys := createStorageAccountOutput.StorageAccountKeys
+
+	logrus.Debugf("StorageAccount.ID=%s", *storageAccount.ID)
+
+	// Create blob storage container
+	createBlobContainerOutput, err := CreateBlobContainer(ctx, &CreateBlobContainerInput{
+		SubscriptionID:       subscriptionID,
+		ResourceGroupName:    resourceGroupName,
+		StorageAccountName:   storageAccountName,
+		ContainerName:        containerName,
+		StorageClientFactory: storageClientFactory,
+	})
+	if err != nil {
+		return err
+	}
+
+	blobContainer := createBlobContainerOutput.BlobContainer
+	logrus.Debugf("BlobContainer.ID=%s", *blobContainer.ID)
+
+	// Upload the image to the container
```

Comment:
This takes so long, I wonder if we should wrap this in an envvar that allows us to skip upload (only for use in development, to speed up iteration) and using marketplace images instead?